sklearn/metrics/_regression.py

```diff
@@ -395,21 +395,29 @@ def mean_absolute_percentage_error(
     >>> mean_absolute_percentage_error(y_true, y_pred)
     112589990684262.48
     """
+    input_arrays = [y_true, y_pred, sample_weight, multioutput]
+    xp, _ = get_namespace(*input_arrays)
+
     y_type, y_true, y_pred, multioutput = _check_reg_targets(
         y_true, y_pred, multioutput
     )
     check_consistent_length(y_true, y_pred, sample_weight)
-    epsilon = np.finfo(np.float64).eps
-    mape = np.abs(y_pred - y_true) / np.maximum(np.abs(y_true), epsilon)
-    output_errors = np.average(mape, weights=sample_weight, axis=0)
+    epsilon = xp.asarray(xp.finfo(xp.float64).eps, dtype=xp.float64)
+    y_true_abs = xp.asarray(xp.abs(y_true), dtype=xp.float64)
+    mape = xp.asarray(xp.abs(y_pred - y_true), dtype=xp.float64) / xp.where(
+        epsilon < y_true_abs, y_true_abs, epsilon
+    )
```

Comment:
Good catch. The original code did not upcast to `float64` so we should not either.

Reply:
Thank you thank you! I also noticed the CUDA failures and I could not fathom why that is apart from #29269. It will make the suggested code changes and hopefully everything passes after #29321 is merged. (PS. everything has always been passing locally)

Reply:
> I might have missed something, why do we need cast to float64 each time we run `abs`?

btw the casting is for `xp.asarray`, not `abs`